File: wrappers/getCombinedLandsatSentinel2Wrapper.js
/**** Start of imports. If edited, may not auto-convert in the playground. ****/
var geometry = /* color: #d63000 */ee.Geometry.Polygon(
        [[[-115.19522231507318, 48.704101580291294],
          [-115.12930434632318, 48.39866293818322],
          [-114.10757583069818, 48.62065285721024],
          [-114.40969985413568, 48.87781053836135]]]);
/***** End of imports. If edited, may not auto-convert in the playground. *****/
//Module imports
var getImageLib = require('users/USFS_GTAC/modules:getImagesLib.js');
///////////////////////////////////////////////////////////////////////////////
// Define user parameters:

// 1. Specify study area: Study area
// Can specify a country, provide a fusion table  or asset table (must add 
// .geometry() after it), or draw a polygon and make studyArea = drawnPolygon
var studyArea = geometry;

// 2. Update the startJulian and endJulian variables to indicate your seasonal 
// constraints. This supports wrapping for tropics and southern hemisphere.
// startJulian: Starting Julian date 
// endJulian: Ending Julian date
var startJulian =190;
var endJulian = 250; 

// 3. Specify start and end years for all analyses
// More than a 3 year span should be provided for time series methods to work 
// well. If using Fmask as the cloud/cloud shadow masking method, this does not 
// matter
var startYear = 2016;
var endYear = 2018;

// 4. Specify an annual buffer to include imagery from the same season 
// timeframe from the prior and following year. timeBuffer = 1 will result 
// in a 3 year moving window
// var timebuffer = 1;

// 5. Specify the weights to be used for the moving window created by timeBuffer
//For example- if timeBuffer is 1, that is a 3 year moving window
//If the center year is 2000, then the years are 1999,2000, and 2001
//In order to overweight the center year, you could specify the weights as
//[1,5,1] which would duplicate the center year 5 times and increase its weight for
//the compositing method
// var weights = [1,5,1];



// 6. Choose medoid or median compositing method. 
// Median tends to be smoother, while medoid retains 
// single date of observation across all bands
// If not exporting indices with composites to save space, medoid should be used
var compositingMethod = 'medoid';

// 7. Choose Top of Atmospheric (TOA) or Surface Reflectance (SR) 
// Specify TOA or SR
// Current implementation does not support Fmask for TOA
var toaOrSR = 'SR';

// 8. Choose whether to include Landat 7
// Generally only included when data are limited
var includeSLCOffL7 = false;

//9. Whether to defringe L5
//Landsat 5 data has fringes on the edges that can introduce anomalies into 
//the analysis.  This method removes them, but is somewhat computationally expensive
var defringeL5 = false;

// 10. Choose cloud/cloud shadow masking method
// Choices are a series of booleans for cloudScore, TDOM, and elements of Fmask
//Fmask masking options will run fastest since they're precomputed
//CloudScore runs pretty quickly, but does look at the time series to find areas that 
//always have a high cloudScore to reduce comission errors- this takes some time
//and needs a longer time series (>5 years or so)
//TDOM also looks at the time series and will need a longer time series
// var applyCloudScore = true;
// var applyFmaskCloudMask = false;

// var applyTDOM = true;
// var applyFmaskCloudShadowMask = false;

// var applyFmaskSnowMask = false;

// 11. Cloud and cloud shadow masking parameters.
// If cloudScoreTDOM is chosen
// cloudScoreThresh: If using the cloudScoreTDOMShift method-Threshold for cloud 
//    masking (lower number masks more clouds.  Between 10 and 30 generally 
//    works best)
var cloudScoreThresh = 10;

//Whether to find if an area typically has a high cloudScore
//If an area is always cloudy, this will result in cloud masking omission
//For bright areas that may always have a high cloudScore
//but not actually be cloudy, this will result in a reduction of commission errors
//This procedure needs at least 5 years of data to work well
var performCloudScoreOffset = false;

// If performCloudScoreOffset = true:
//Percentile of cloud score to pull from time series to represent a minimum for 
// the cloud score over time for a given pixel. Reduces comission errors over 
// cool bright surfaces. Generally between 5 and 10 works well. 0 generally is a
// bit noisy but may be necessary in persistently cloudy areas
var cloudScorePctl = 10;

//Sentinel 2 only!
//Height of clouds to use to project cloud shadows
var cloudHeights = ee.List.sequence(500,10000,500);

// zScoreThresh: Threshold for cloud shadow masking- lower number masks out 
//    less. Between -0.8 and -1.2 generally works well
var zScoreThresh = -1;

// shadowSumThresh: Sum of IR bands to include as shadows within TDOM and the 
//    shadow shift method (lower number masks out less)
var shadowSumThresh = 0.35;

// contractPixels: The radius of the number of pixels to contract (negative 
//    buffer) clouds and cloud shadows by. Intended to eliminate smaller cloud 
//    patches that are likely errors
// (1.5 results in a -1 pixel buffer)(0.5 results in a -0 pixel buffer)
// (1.5 or 2.5 generally is sufficient)
var contractPixels = 1.5; 

// dilatePixels: The radius of the number of pixels to dilate (buffer) clouds 
//    and cloud shadows by. Intended to include edges of clouds/cloud shadows 
//    that are often missed
// (1.5 results in a 1 pixel buffer)(0.5 results in a 0 pixel buffer)
// (2.5 or 3.5 generally is sufficient)
var dilatePixels = 3.5;

// // 12. correctIllumination: Choose if you want to correct the illumination using
// // Sun-Canopy-Sensor+C correction. Additionally, choose the scale at which the
// // correction is calculated in meters.
// var correctIllumination = false;
// var correctScale = 250;//Choose a scale to reduce on- 250 generally works well

// //13. Export params
// //Whether to export composites
// var exportComposites = true;

//Set up Names for the export
var outputName = 'Landsat';

//Provide location composites will be exported to
//This should be an asset folder, or more ideally, an asset imageCollection
var exportPathRoot = 'users/ianhousman/test/changeCollection';



//CRS- must be provided.  
//Common crs codes: Web mercator is EPSG:4326, USGS Albers is EPSG:5070, 
//WGS84 UTM N hemisphere is EPSG:326+ zone number (zone 12 N would be EPSG:32612) and S hemisphere is EPSG:327+ zone number
var crs = 'EPSG:32717'//'EPSG:5070';

//Specify transform if scale is null and snapping to known grid is needed
var transform = null;//[30,0,-2361915.0,0,-30,3177735.0];

//Specify scale if transform is null
var scale = 20;


///////////////////////////////////////////////////////////////////////
// End user parameters
///////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////
//Start function calls

// Prepare dates
//Wrap the dates if needed
var wrapOffset = 0;
// if (startJulian > endJulian) {
//   wrapOffset = 365;
// }
var startDate = ee.Date.fromYMD(startYear,1,1).advance(startJulian-1,'day');
var endDate = ee.Date.fromYMD(endYear,1,1).advance(endJulian-1+wrapOffset,'day');
print('Start and end dates:', startDate, endDate);

////////////////////////////////////////////////////////////////////////////////
//Get Landsat and Sentinel 2 raw images
var ls = getImageLib.getImageCollection(studyArea,startDate,endDate,startJulian,endJulian,
    toaOrSR,includeSLCOffL7,defringeL5);
var s2s = getImageLib.getS2(studyArea,startDate,endDate,startJulian,endJulian);
Map.addLayer(ls.first(),getImageLib.vizParamsFalse,'Landsat No Masking',false)

//Apply respective cloudScore functions
ls = getImageLib.applyCloudScoreAlgorithm(ls,getImageLib.landsatCloudScore,cloudScoreThresh,cloudScorePctl,contractPixels,dilatePixels,performCloudScoreOffset);
s2s = getImageLib.applyCloudScoreAlgorithm(s2s,getImageLib.sentinel2CloudScore,cloudScoreThresh,cloudScorePctl,contractPixels,dilatePixels,performCloudScoreOffset);
Map.addLayer(ls.first(),getImageLib.vizParamsFalse,'Landsat Cloud Masking',false)
//Set a property for splitting apart later
ls = ls.map(function(img){return img.float().set('whichProgram','Landsat')});
s2s = s2s.map(function(img){return img.float().set('whichProgram','Sentinel2')});

//Merge collections
var merged = ls.merge(s2s);

//Perform TDOM
merged = getImageLib.simpleTDOM2(merged,zScoreThresh,shadowSumThresh,contractPixels,dilatePixels);

//Seperate back out
ls = merged.filter(ee.Filter.eq('whichProgram','Landsat'));
s2s = merged.filter(ee.Filter.eq('whichProgram','Sentinel2'));
Map.addLayer(ls.first(),getImageLib.vizParamsFalse,'Landsat Cloud/Shadow Masking',false)
// // Create composite time series function
// function createAndExportComposites(c,startYear,endYear,startJulian,endJulian,timebuffer,weights,everyHowManyDays,exportPathRoot,exportName,exportBands,nonDivideBands,scale,crs,transform){

// //Iterate across each year
// ee.List.sequence(startYear+timebuffer,endYear-timebuffer).getInfo().map(function(year){
//     var dummyImage = ee.Image(c.first());
//     // Set up dates
//     var startYearT = year-timebuffer;
//     var endYearT = year+timebuffer;
//   //Set up weighted moving widow
//     var yearsT = ee.List.sequence(startYearT,endYearT);
    
//     var z = yearsT.zip(weights);
//     var yearsTT = z.map(function(i){
//       i = ee.List(i);
//       return ee.List.repeat(i.get(0),i.get(1));
//     }).flatten();
    
//   ee.List.sequence(startJulian,endJulian,everyHowManyDays).getInfo().map(function(startJulianT){
//     var endJulianT = startJulianT+everyHowManyDays-1;

//     if(endJulianT <= endJulian){
//       // print(startYearT,endYearT,year,startJulianT,endJulianT);
//       //Iterate across each year in list
//       var images = yearsTT.map(function(yrT){
       
        
//         // Filter images for given date range
//         var cT = c.filter(ee.Filter.calendarRange(yrT,yrT,'year'))
//                     .filter(ee.Filter.calendarRange(startJulianT,endJulianT));
//         cT = getImageLib.fillEmptyCollections(cT,dummyImage);
//         return cT;
//       });
//       var cT = ee.ImageCollection(ee.FeatureCollection(images).flatten());
//       var count = cT.select([0]).count().rename(['count'])
//       // Compute median or medoid or apply reducer
//     var composite;
//     if (compositingMethod.toLowerCase() === 'median') {
//       composite = cT.median();
//     }
//     else {
      
//       composite = getImageLib.medoidMosaicMSD(cT,['blue','green','red','nir','swir1','swir2']);
//     }
//     composite = composite.addBands(count).select(exportBands);
    
    
    
//     // Reformat data for export
//     var compositeBands = composite.bandNames();
//     if(nonDivideBands != null){
//       var composite10k = composite.select(compositeBands.removeAll(nonDivideBands))
//       .multiply(10000);
//       composite = composite10k.addBands(composite.select(nonDivideBands))
//       .select(compositeBands).int16();
//     }
//     else{
//       composite = composite.multiply(10000).int16();
//     }
    
    
//     var startDate = ee.Date.fromYMD(year,1,1).advance(startJulianT,'day').millis();
//     var endDate = ee.Date.fromYMD(year,1,1).advance(endJulianT,'day').millis();
    
//     composite = composite.set({
//                         'system:time_start':startDate,
//                         'system:time_end':endDate,
//                         'startJulian':startJulian,
//                         'endJulian':endJulian,
//                         'yearBuffer':timebuffer,
//                         'yearWeights': getImageLib.listToString(weights),
//                         'yrCenter':year,
//                         'dilatePixels':dilatePixels,
//                         'contractPixels':contractPixels,
//                         'cloudScoreThresh':cloudScoreThresh,
//                         'useCloudScore':true,
//                         'crs':crs,
//                         'startDOY':startJulianT,
//                         'endDOY':endJulianT,
//                         'useSRmask':false,
//                         'shadowSumThresh':shadowSumThresh,
//                         'zScoreThresh':zScoreThresh,
//                         'terrain':false.toString(),
//                         'useCloudProject':performCloudScoreOffset.toString(),
//                         'useTDOM':true.toString(),
//                         'useLandsatS2HybridTDOM':true.toString(),
//                         'whichProgram':exportName,
//                         'compositingMethod':compositingMethod.toLowerCase()
                     
//     });
    


//     var outName = exportName+'_y'+startYearT.toString() + '_'+ endYearT.toString() + '_j'+startJulianT.toString() + '_' + endJulianT.toString();
//     // Map.addLayer(composite,{min:500,max:5000,bands:'swir1,nir,red'},outName,false);
    
    
//     var exportPath = exportPathRoot + '/' + outName;
//     // print('Write down the Asset ID:', exportPath);
  
//     getImageLib.exportToAssetWrapper(composite,outName,exportPath,'mean',
//       studyArea,scale,crs,transform);
    
//     }
// });
// });
// }

// //Export bands for each program
// var lExportBands = [ 'blue', 'green', 'red','nir','swir1', 'swir2','temp','count'];
// var S2ExportBands = ['cb', 'blue', 'green', 'red', 're1','re2','re3','nir', 'nir2', 'waterVapor', 'cirrus','swir1', 'swir2','count'];

// //Export bi-weekly
// createAndExportComposites(ls,2000,2018,1,365,0,[1],14,'projects/Sacha/L8/L8_Biweekly_Updated','Landsat',lExportBands,['temp','count'],scale,crs,null);
// createAndExportComposites(s2s,2015,2018,1,365,0,[1],14,'projects/Sacha/S2/S2_Biweekly_Updated','Sentinel2',S2ExportBands,['count'],scale,crs,null);

//Export annuals
// createAndExportComposites(ls,1999,2019,1,365,1,[1,3,1],365,'projects/Sacha/L8/L8_Annual_Updated','Landsat',lExportBands,['temp','count'],scale,crs,null);
// createAndExportComposites(s2s,2014,2019,1,365,1,[1,3,1],365,'projects/Sacha/S2/S2_Annual_Updated','Sentinel2',S2ExportBands,['count'],scale,crs,null);

/////////////////////////////////////////////////////
//Code for starting all tasks once this script has ran
//Press f12, then paste functions into console
//Then paste function calls into console
// function runTaskList() {


//     //1. task local type-EXPORT_FEATURES awaiting-user-config

//     //2. task local type-EXPORT_IMAGE awaiting-user-config

//     var tasklist = document.getElementsByClassName('awaiting-user-config');

//     for (var i = 0; i < tasklist.length; i++)

//         tasklist[i].children[2].click();

// }

// // confirmAll();

// function confirmAll() {

//     var ok = document.getElementsByClassName('goog-buttonset-default goog-buttonset-action');

//     for (var i = 0; i < ok.length; i++)

//         ok[i].click();

// }

File: examples/getCombinedLandsatSentinel2Wrapper.py
"""
   Copyright 2021 Ian Housman

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

#Example of how to get Landsat data using the getImagesLib and view outputs using the Python visualization tools
#Acquires Landsat data and then adds them to the viewer
####################################################################################################
import os,sys
sys.path.append(os.getcwd())

#Module imports
import geeViz.getImagesLib as getImagesLib
ee = getImagesLib.ee
Map = getImagesLib.Map
Map.clearMap()
####################################################################################################
####################################################################################################
#Define user parameters:

# Specify study area: Study area
# Can be a featureCollection, feature, or geometry
studyArea = getImagesLib.testAreas['CA']

# Update the startJulian and endJulian variables to indicate your seasonal 
# constraints. This supports wrapping for tropics and southern hemisphere.
# If using wrapping and the majority of the days occur in the second year, the system:time_start will default 
# to June 1 of that year.Otherwise, all system:time_starts will default to June 1 of the given year
# startJulian: Starting Julian date 
# endJulian: Ending Julian date
startJulian = 152
endJulian = 273

# Specify start and end years for all analyses
# More than a 3 year span should be provided for time series methods to work 
# well. If providing pre-computed stats for cloudScore and TDOM, this does not 
# matter
startYear = 2018
endYear = 2021

# Specify an annual buffer to include imagery from the same season 
# timeframe from the prior and following year. timeBuffer = 1 will result 
# in a 3 year moving window. If you want single-year composites, set to 0
timebuffer =0

# Specify the weights to be used for the moving window created by timeBuffer
# For example- if timeBuffer is 1, that is a 3 year moving window
# If the center year is 2000, then the years are 1999,2000, and 2001
# In order to overweight the center year, you could specify the weights as
# [1,5,1] which would duplicate the center year 5 times and increase its weight for
# the compositing method. If timeBuffer = 0, set to [1]
weights = [1]

# Choose medoid or median compositing method. 
# Median tends to be smoother, while medoid retains 
# single date of observation across all bands
# The date of each pixel is stored if medoid is used. This is not done for median
# If not exporting indices with composites to save space, medoid should be used
compositingMethod = 'medoid'

# Choose Top of Atmospheric (TOA) or Surface Reflectance (SR)
# Use caution when combining Landsat and S2 SR data since S2 SR data had terrain correction performed 
toaOrSR = 'TOA'

# Whether to convert S2 images from the military grid reference system(MGRS) tiles to daily mosaics to avoid arbitrary
# MGRS tile artifacts or not. In most cases, it is best to set this to true.
convertToDailyMosaics = True


# Choose whether to include Landat 7
# Generally only included when data are limited
includeSLCOffL7 = True

# Whether to defringe L4 and L5
# Landsat 4 and 5 data have fringes on the edges that can introduce anomalies into 
# the analysis.  This method removes them, but is somewhat computationally expensive
defringeL5 = True

# Choose cloud/cloud shadow masking method
#  Choices are a series of booleans for cloudScore, TDOM, and elements of Fmask
# Fmask masking options will run fastest since they're precomputed
# Fmask cloud mask is generally very good, while the fMask cloud shadow
# mask isn't great. TDOM tends to perform better than the Fmask cloud shadow mask. cloudScore 
# is usually about as good as the Fmask cloud mask overall, but each failes in different instances.
# CloudScore runs pretty quickly, but does look at the time series to find areas that 
# always have a high cloudScore to reduce commission errors- this takes some time
# and needs a longer time series (>5 years or so)
# TDOM also looks at the time series and will need a longer time series
# If pre-computed cloudScore offsets and/or TDOM stats are provided below, cloudScore
# and TDOM will run quite quickly

#CloudScore and TDOM switches- for both Sentinel 2 and Landsat separately
#We generally use cloudScore for Landsat, but not Sentinel 2
applyCloudScoreLandsat = True
applyCloudScoreSentinel2 = False

applyTDOMLandsat = True
applyTDOMSentinel2 = True

#S2 only cloud/cloud shadow masking methods switches- generally do not use these
#QA band method is fast but is generally awful- don't use if you like good composites
#Shadow shift is intended if you don't have a time series to use for TDOM or just want individual images
#It will commit any dark area that the cloud mask is cast over (water, hill shadows, etc)
applyQABand = False
applyShadowShift = False
#Height of clouds to use to project cloud shadows
cloudHeights = ee.List.sequence(500,10000,500)

#Whether to use the pre-computed cloud probabilities to mask
#clouds for Sentinel 2
#This method works really well
applyCloudProbability = True

#If cloudProbability is chosen, choose a threshold 
#(generally somewhere around 40-60 works well)
cloudProbThresh = 40

#Fmask switches- only for Landsat
#Generally we do use these
applyFmaskCloudMask = True
applyFmaskCloudShadowMask = True
applyFmaskSnowMask = False

# Cloud and cloud shadow masking parameters.
# If cloudScore  is chosen
# cloudScoreThresh: If using the cloudScore or cloudProbability method-Threshold for cloud 
#    masking (lower number masks more clouds.  Between 10 and 30 generally 
#    works best)
cloudScoreThresh = 20

#Whether to find if an area typically has a high cloudScore
#If an area is always cloudy, this will result in cloud masking omission
#For bright areas that may always have a high cloudScore
#but not actually be cloudy, this will result in a reduction of commission errors
#This procedure needs at least 5 years of data to work well
performCloudScoreOffset = True

# If performCloudScoreOffset = true:
#Percentile of cloud score to pull from time series to represent a minimum for 
# the cloud score over time for a given pixel. Reduces comission errors over 
# cool bright surfaces. Generally between 5 and 10 works well. 0 generally is a
# bit noisy but may be necessary in persistently cloudy areas
cloudScorePctl = 10

# zScoreThresh: Threshold for cloud shadow masking- lower number masks out 
#    less. Between -0.8 and -1.2 generally works well
zScoreThresh = -1

# shadowSumThresh: Sum of IR bands to include as shadows within TDOM and the 
#    shadow shift method (lower number masks out less)
shadowSumThresh = 0.35

# contractPixels: The radius of the number of pixels to contract (negative 
#    buffer) clouds and cloud shadows by. Intended to eliminate smaller cloud 
#    patches that are likely errors
# (1.5 results in a -1 pixel buffer)(0.5 results in a -0 pixel buffer)
# (1.5 or 2.5 generally is sufficient)
contractPixels = 1.5

# dilatePixels: The radius of the number of pixels to dilate (buffer) clouds 
#    and cloud shadows by. Intended to include edges of clouds/cloud shadows 
#    that are often missed
# (1.5 results in a 1 pixel buffer)(0.5 results in a 0 pixel buffer)
# (2.5 or 3.5 generally is sufficient)
dilatePixels = 2.5

# Choose the resampling method: 'aggregate','near', 'bilinear', or 'bicubic'
# Defaults to 'aggregate' for Sentinel 2 and 'near' for Landsat

# Aggregate is generally useful for aggregating pixels when reprojecting instead of resampling
# A good example would be reprojecting S2 data to 30 m

# If method other than 'near' is chosen, any map drawn on the fly that is not
# reprojected, will appear blurred or not really represented properly
# Use .reproject to view the actual resulting image (this will slow it down)
landsatResampleMethod = 'near'

sentinel2ResampleMethod = 'aggregate'

# Choose whether to use the Chastain et al 2019(https://www.sciencedirect.com/science/article/pii/S0034425718305212)
# harmonization method
# All harmonization models apply a rather small correction and are likely not needed
#Only runs if toaOrSR = 'TOA'
runChastainHarmonization = True

# If available, bring in preComputed cloudScore offsets and TDOM stats
# Set to null if computing on-the-fly is wanted
# These have been pre-computed for all CONUS for Landsat and Setinel 2 (separately)
# and are appropriate to use for any time period within the growing season
# The cloudScore offset is generally some lower percentile of cloudScores on a pixel-wise basis
preComputedCloudScoreOffset = getImagesLib.getPrecomputedCloudScoreOffsets(cloudScorePctl)
preComputedLandsatCloudScoreOffset = preComputedCloudScoreOffset['landsat']
preComputedSentinel2CloudScoreOffset = preComputedCloudScoreOffset['sentinel2']

# The TDOM stats are the mean and standard deviations of the two bands used in TDOM
# By default, TDOM uses the nir and swir1 bands
preComputedTDOMStats = getImagesLib.getPrecomputedTDOMStats()
preComputedLandsatTDOMIRMean = preComputedTDOMStats['landsat']['mean']
preComputedLandsatTDOMIRStdDev = preComputedTDOMStats['landsat']['stdDev']

preComputedSentinel2TDOMIRMean = preComputedTDOMStats['sentinel2']['mean']
preComputedSentinel2TDOMIRStdDev = preComputedTDOMStats['sentinel2']['stdDev']


# Export params
# Whether to export composites
exportComposites = False

#Set up Names for the export
outputName = 'Landsat_Sentinel2_Hybrid'

# Provide location composites will be exported to
# This should be an asset folder, or more ideally, an asset imageCollection
exportPathRoot = 'users/iwhousman/test/compositeCollection'



# CRS- must be provided.  
# Common crs codes: Web mercator is EPSG:4326, USGS Albers is EPSG:5070, 
# WGS84 UTM N hemisphere is EPSG:326+ zone number (zone 12 N would be EPSG:32612) and S hemisphere is EPSG:327+ zone number
crs = 'EPSG:5070'

# Specify transform if scale is null and snapping to known grid is needed
transform = [10,0,-2361915.0,0,-10,3177735.0]

# Specify scale if transform is null
scale = None
####################################################################################################
#End user parameters
####################################################################################################
####################################################################################################
####################################################################################################
#Start function calls
####################################################################################################
####################################################################################################
processedAndComposites = getImagesLib.getLandsatAndSentinel2HybridWrapper(studyArea,startYear,endYear,startJulian,endJulian,
  timebuffer,weights,compositingMethod,toaOrSR,includeSLCOffL7,defringeL5,
  applyQABand,applyCloudProbability,applyShadowShift,applyCloudScoreLandsat,applyCloudScoreSentinel2,applyTDOMLandsat,applyTDOMSentinel2,applyFmaskCloudMask,applyFmaskCloudShadowMask,applyFmaskSnowMask,
  cloudHeights,cloudScoreThresh,performCloudScoreOffset,cloudScorePctl,
  zScoreThresh,shadowSumThresh,contractPixels,dilatePixels,['nir','swir1'],landsatResampleMethod,sentinel2ResampleMethod,convertToDailyMosaics,runChastainHarmonization,
  False,250,
  exportComposites,outputName,exportPathRoot,crs,transform,scale,
  preComputedLandsatCloudScoreOffset,preComputedLandsatTDOMIRMean,preComputedLandsatTDOMIRStdDev,preComputedSentinel2CloudScoreOffset,preComputedSentinel2TDOMIRMean,preComputedSentinel2TDOMIRStdDev,
  cloudProbThresh);

#Separate into scenes and composites for subsequent analysis
processedScenes = processedAndComposites['processedScenes']
processedComposites = processedAndComposites['processedComposites']

# Map.addLayer(processedComposites.select(['NDVI','NBR']),{'addToLegend':'false'},'Time Series (NBR and NDVI)',False)
for year in range(startYear + timebuffer      ,endYear + 1 - timebuffer ):
     t = processedComposites.filter(ee.Filter.calendarRange(year,year,'year')).mosaic()
     Map.addLayer(t.float(),getImagesLib.vizParamsFalse,str(year),False)
####################################################################################################
#Load the study region
Map.addLayer(studyArea, {'strokeColor': '0000FF'}, "Study Area", True)
Map.centerObject(studyArea)
####################################################################################################
####################################################################################################
Map.view()